Enhance validation error handling in ValidationEndpointFilterFactory using IProblemDetailsService

marcominerva opened the pull request:
# Modify validation error handling to utilize IProblemDetailsService for response writing

- [X] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [X] You've included unit or integration tests for your change, where applicable.
- [X] You've included inline docs for your change, where applicable.
- [X] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

This PR updates the `ValidationEndpointFilterFactory` to improve how HTTP validation errors are handled, specifically by integrating with the `IProblemDetailsService` for enhanced response customization. The changes include adjustments to the response logic and dependency usage.

## Description

* **Integration with `IProblemDetailsService`**: Added logic to check for and utilize `IProblemDetailsService` to write custom problem details responses if available. This allows for more flexible and centralized error handling.
* **Fallback to default implementation**: Ensured a fallback mechanism to return `HttpValidationProblemDetails` if `IProblemDetailsService` is not present.

Fixes #61219

marcominerva committed:
Enhance validation error handling in filter factory

Modify validation error handling to utilize IProblemDetailsService for response writing.
Fallback to default behavior if the service is unavailable.

martincostello commented:
I imagine the team will want a test for this.

marcominerva commented:
At the moment, there are no tests at all for the `ValidationEndpointFilterFactory` class. I need to figure out the correct way to add them, but in the meantime, I decided to open the PR because I’d like to understand whether this is the right approach.
Specifically, I’m referring to returning `EmptyHttpResult.Instance`: I’ve verified that it works, but I’m wondering if there’s a better solution.

In the meantime, I’ll think about adding a test.

martincostello commented:
Just suggested it as it wasn't a draft 😃

marcominerva commented:
You're right, @martincostello — I just converted the PR to a draft until the tests are done. 😉

marcominerva committed:
Add validation filter tests for ProblemDetails handling

Introduces the `ValidationFilterEndpointFactoryTests` class with two unit tests. These tests validate HTTP request behavior with and without the `ProblemDetails` service registered, ensuring correct status codes, content types, and the structure of the returned `ProblemDetails` object for validation errors.

marcominerva committed:
Fix a typo

Rename validation test class for clarity

marcominerva commented:
I finished earlier than expected. I’ve added a couple of tests, so the PR is now ready for review.

captainsafia commented:
@marcominerva Thanks for taking the time to working on this! Sorry I never got back to your question on the issue.

> At the moment, there are no tests at all for the ValidationEndpointFilterFactory class. I need to figure out the correct way to add them, but in the meantime, I decided to open the PR because I’d like to understand whether this is the right approach.
Specifically, I’m referring to returning EmptyHttpResult.Instance: I’ve verified that it works, but I’m wondering if there’s a better solution.

There's coverage for the filter in the ValidationsGenerator tests which execute the end-to-end but adding individual tests is good.

One thing to note is that since we don't have a runtime type resolver (see https://github.com/dotnet/aspnetcore/issues/61220), we won't be able to test validation against complex types but that might be fine for the purpose of this test.

captainsafia commented on src/Http/Routing/src/ValidationEndpointFilterFactory.cs:
```diff
@@ -87,8 +89,27 @@ public static EndpointFilterDelegate Create(EndpointFilterFactoryContext context
+                        ProblemDetails = problemDetails
+                    }))
+                    {
+                        // We need to prevent further execution, because the actual
```
I think with this pattern, we've lost the ability for users to add endpoint filters after this one that can inspect what the validation errors that were returned were. Since we lack `ModelState` in minimal APIs, this might be the only way for users to understand the error state.

marcominerva replied:
In fact, this is the point I mentioned I was unsure about.

However, I have verified that, even with the current implementation, custom endpoint filters are not invoked when validation fails, or am I missing something?

captainsafia replied:
You're correct -- I got my wires crossed. Our only support for circumventing the default filter at the moment is disabling it and adding your own explicit one. In that context, this approach seems fine.

captainsafia commented on src/Http/Http.Extensions/test/ValidationFilterEndpointFactoryTests.cs:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    [Fact]
+    public async Task UseProblemDetailsServiceWhenAddedInServiceCollection()
```
Can we add a test case that covers calling `AddProblemDetails` with a callback and verifying the customizations that are applied in that callback register?

marcominerva committed:
Enhance validation endpoint tests

Added a new test method to validate the ProblemDetails service with a timestamp extension.

captainsafia committed:
Apply suggestions from code review
